File: native/strategy_core/src/minutetemp.rs
use std::collections::BTreeMap;

use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Deserializer, Serialize, de};

pub type ReportType = String;
pub type TemperatureUnit = String;
pub type PlanTier = String;
pub type DataResolution = String;
pub type TemperatureDayMode = String;
pub type WuDayMode = String;
pub type OracleScoreMode = String;
pub type OracleRankBy = String;
pub type ReportScheduleBasis = String;

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct CityInfo {
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub slug: String,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub timezone: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct StationInfo {
    #[serde(default)]
    pub station_id: String,
    #[serde(default)]
    pub name: String,
    #[serde(default = "default_temperature_unit")]
    pub temperature_unit: String,
    pub uses_nws_climate_day: Option<bool>,
}

impl Default for StationInfo {
    fn default() -> Self {
        Self {
            station_id: String::new(),
            name: String::new(),
            temperature_unit: default_temperature_unit(),
            uses_nws_climate_day: None,
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct ObservationRecord {
    pub observation_time: Option<DateTime<Utc>>,
    pub temperature_f: Option<f64>,
    pub temperature_c: Option<f64>,
    pub dewpoint: Option<f64>,
    pub heat_index: Option<f64>,
    pub wind_chill: Option<f64>,
    pub relative_humidity: Option<f64>,
    pub barometric_pressure: Option<f64>,
    pub sea_level_pressure: Option<f64>,
    pub wind_speed: Option<f64>,
    pub wind_direction: Option<f64>,
    pub wind_gust: Option<f64>,
    pub text_description: Option<String>,
    pub precipitation_1h: Option<f64>,
    pub precipitation_3h: Option<f64>,
    pub precipitation_6h: Option<f64>,
    #[serde(default)]
    pub is_locf: bool,
    #[serde(default)]
    pub is_from_report: bool,
    pub report_type: Option<String>,
    pub source_report_id: Option<String>,
    pub temp_min_f: Option<f64>,
    pub temp_max_f: Option<f64>,
    pub temp_min_c: Option<f64>,
    pub temp_max_c: Option<f64>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct StationReportRecord {
    #[serde(default)]
    pub report_id: String,
    #[serde(default)]
    pub report_revision: i64,
    pub report_updated_at: Option<DateTime<Utc>>,
    pub report_type: Option<String>,
    pub report_date: Option<NaiveDate>,
    pub issuance_time: Option<DateTime<Utc>>,
    pub fetched_at: Option<DateTime<Utc>>,
    pub max_temp_f: Option<f64>,
    pub max_temp_c: Option<f64>,
    pub max_temp_time_utc: Option<DateTime<Utc>>,
    pub min_temp_f: Option<f64>,
    pub min_temp_c: Option<f64>,
    pub min_temp_time_utc: Option<DateTime<Utc>>,
    pub temp_f: Option<f64>,
    pub temp_c: Option<f64>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ReportClockSchedule {
    #[serde(default = "default_report_schedule_basis")]
    pub basis: String,
    pub hour: Option<i64>,
    pub minute: Option<i64>,
    pub utc_hour: Option<i64>,
    pub utc_minute: Option<i64>,
    pub local_hour: Option<i64>,
    pub local_minute: Option<i64>,
    #[serde(default)]
    pub label: String,
}

impl Default for ReportClockSchedule {
    fn default() -> Self {
        Self {
            basis: default_report_schedule_basis(),
            hour: None,
            minute: None,
            utc_hour: None,
            utc_minute: None,
            local_hour: None,
            local_minute: None,
            label: String::new(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ReportIntervalSchedule {
    pub interval_minutes: i64,
    pub utc_minute: Option<i64>,
    pub local_minute: Option<i64>,
    #[serde(default)]
    pub label: String,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct ReportMultiHourSchedule {
    #[serde(default)]
    pub utc_hours: Vec<i64>,
    #[serde(default)]
    pub local_hours: Vec<i64>,
    pub utc_minute: Option<i64>,
    pub local_minute: Option<i64>,
    #[serde(default)]
    pub label: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
#[serde(untagged)]
pub enum ReportSchedule {
    Clock(ReportClockSchedule),
    Interval(ReportIntervalSchedule),
    MultiHour(ReportMultiHourSchedule),
}

impl<'de> Deserialize<'de> for ReportSchedule {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let value = serde_json::Value::deserialize(deserializer)?;
        let object = value
            .as_object()
            .ok_or_else(|| de::Error::custom("report schedule must be an object"))?;

        if object.contains_key("interval_minutes") {
            return serde_json::from_value(value)
                .map(Self::Interval)
                .map_err(de::Error::custom);
        }
        if object.contains_key("utc_hours") || object.contains_key("local_hours") {
            return serde_json::from_value(value)
                .map(Self::MultiHour)
                .map_err(de::Error::custom);
        }
        serde_json::from_value(value)
            .map(Self::Clock)
            .map_err(de::Error::custom)
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ReportScheduleEntry {
    Single(ReportSchedule),
    Multiple(Vec<ReportSchedule>),
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct HourlyForecastRecord {
    pub time: Option<DateTime<Utc>>,
    pub temperature_2m_f: Option<f64>,
    pub temperature_2m_c: Option<f64>,
    pub apparent_temperature_f: Option<f64>,
    pub relative_humidity_2m: Option<f64>,
    pub dew_point_2m: Option<f64>,
    pub pressure_msl: Option<f64>,
    pub wind_speed_10m: Option<f64>,
    pub wind_direction_10m: Option<f64>,
    pub wind_gusts_10m: Option<f64>,
    pub cloud_cover: Option<f64>,
    pub precipitation_probability: Option<f64>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct ForecastBundleRun {
    #[serde(default)]
    pub id: String,
    pub fetched_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub timezone: String,
    pub utc_offset_seconds: Option<i64>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct ForecastBundle {
    #[serde(default)]
    pub model_id: String,
    pub forecast_run: Option<ForecastBundleRun>,
    #[serde(default)]
    pub hourly: Vec<HourlyForecastRecord>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct OracleModelScoreRecord {
    #[serde(default)]
    pub model_id: String,
    #[serde(default)]
    pub model_name: String,
    pub is_public: Option<bool>,
    pub high_mae: Option<f64>,
    pub low_mae: Option<f64>,
    pub high_bias: Option<f64>,
    pub low_bias: Option<f64>,
    pub combined_mae: Option<f64>,
    pub day_count: Option<i64>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct OracleScoreData {
    #[serde(default)]
    pub station_id: String,
    pub range_start: Option<NaiveDate>,
    pub range_end: Option<NaiveDate>,
    pub days_requested: Option<i64>,
    pub all_time: Option<bool>,
    #[serde(default = "default_oracle_score_mode")]
    pub score_mode: String,
    #[serde(default = "default_oracle_rank_by")]
    pub rank_by: String,
    #[serde(default)]
    pub scores: Vec<OracleModelScoreRecord>,
}

impl Default for OracleScoreData {
    fn default() -> Self {
        Self {
            station_id: String::new(),
            range_start: None,
            range_end: None,
            days_requested: None,
            all_time: None,
            score_mode: default_oracle_score_mode(),
            rank_by: default_oracle_rank_by(),
            scores: Vec::new(),
        }
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct CursorPage {
    pub limit: Option<i64>,
    pub next_cursor: Option<String>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct ForecastRunSummary {
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub station_id: String,
    #[serde(default)]
    pub model_id: String,
    pub forecast_time: Option<DateTime<Utc>>,
    pub fetched_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub timezone: String,
    pub utc_offset_seconds: Option<i64>,
    #[serde(default)]
    pub data_hash: String,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct IpGuardLimits {
    pub requests_per_second: Option<i64>,
    pub burst: Option<i64>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct EffectiveLimits {
    #[serde(default = "default_plan_tier")]
    pub tier: String,
    pub requests_per_minute: Option<i64>,
    pub daily_max: Option<i64>,
    pub max_history_days: Option<i64>,
    pub ip_guard: Option<IpGuardLimits>,
    pub rate_limit_remaining: Option<i64>,
    pub rate_limit_reset_seconds: Option<i64>,
}

impl Default for EffectiveLimits {
    fn default() -> Self {
        Self {
            tier: default_plan_tier(),
            requests_per_minute: None,
            daily_max: None,
            max_history_days: None,
            ip_guard: None,
            rate_limit_remaining: None,
            rate_limit_reset_seconds: None,
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct LatestObservationData {
    pub city: Option<CityInfo>,
    pub station: Option<StationInfo>,
    pub observation: Option<ObservationRecord>,
    pub daily_high_f: Option<f64>,
    pub daily_low_f: Option<f64>,
    pub daily_high_c: Option<f64>,
    pub daily_low_c: Option<f64>,
    pub asos_daily_high_f: Option<f64>,
    pub asos_daily_low_f: Option<f64>,
    pub asos_daily_high_c: Option<f64>,
    pub asos_daily_low_c: Option<f64>,
    pub wu_current_temp_f: Option<f64>,
    pub wu_current_temp_c: Option<f64>,
    pub wu_daily_high_f: Option<f64>,
    pub wu_daily_low_f: Option<f64>,
    pub wu_daily_high_c: Option<f64>,
    pub wu_daily_low_c: Option<f64>,
    pub wu_observation_time: Option<DateTime<Utc>>,
    pub wu_fetched_at: Option<DateTime<Utc>>,
    pub temperature_day_mode: Option<String>,
    pub temperature_day_date: Option<NaiveDate>,
    pub wu_day_mode: Option<String>,
    pub wu_day_date: Option<NaiveDate>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct StationForecastData {
    pub city: Option<CityInfo>,
    pub station: Option<StationInfo>,
    #[serde(default)]
    pub forecasts: Vec<ForecastBundle>,
    pub count: Option<i64>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct LatestReportsData {
    #[serde(default)]
    pub reports: Vec<StationReportRecord>,
    pub report_schedules: Option<BTreeMap<String, ReportScheduleEntry>>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct StationReportsData {
    #[serde(default)]
    pub reports: Vec<StationReportRecord>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct StationReportHistoryPage {
    pub city: Option<CityInfo>,
    pub station: Option<StationInfo>,
    #[serde(default)]
    pub reports: Vec<StationReportRecord>,
    pub count: Option<i64>,
    pub page: Option<CursorPage>,
    pub report_schedules: Option<BTreeMap<String, ReportScheduleEntry>>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct ForecastRunsPage {
    pub city: Option<CityInfo>,
    pub station: Option<StationInfo>,
    #[serde(default)]
    pub runs: Vec<ForecastRunSummary>,
    pub count: Option<i64>,
    pub page: Option<CursorPage>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct ForecastRunData {
    pub city: Option<CityInfo>,
    pub station: Option<StationInfo>,
    pub forecast_run: Option<ForecastRunSummary>,
    #[serde(default)]
    pub hourly: Vec<HourlyForecastRecord>,
    pub count: Option<i64>,
}

fn default_temperature_unit() -> String {
    "F".to_string()
}

fn default_report_schedule_basis() -> String {
    "utc".to_string()
}

fn default_oracle_score_mode() -> String {
    "overall".to_string()
}

fn default_oracle_rank_by() -> String {
    "combined".to_string()
}

fn default_plan_tier() -> String {
    "starter".to_string()
}

File: strategy_core/minutetemp.py
"""MinuteTemp OpenAPI-aligned models for engine-owned data reads."""

from __future__ import annotations

from dataclasses import dataclass, field
from types import MappingProxyType
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from collections.abc import Iterable, Mapping
    from datetime import date, datetime

ReportType = Literal["cli", "dsm", "metar_tgroup", "metar_6hr"]
TemperatureUnit = Literal["F", "C"]
PlanTier = Literal["starter", "pro", "clanker"]
DataResolution = Literal["1m", "5m", "10m"]
TemperatureDayMode = Literal["calendar_day", "nws_climate_day"]
WuDayMode = Literal["calendar_day"]
OracleScoreMode = Literal["overall", "day_ahead", "day_of"]
OracleRankBy = Literal["combined", "high", "low"]
ReportScheduleBasis = Literal["utc", "local"]


def _freeze_tuple[T](values: Iterable[T]) -> tuple[T, ...]:
    return tuple(values)


def _freeze_mapping[K, V](values: Mapping[K, V]) -> Mapping[K, V]:
    return MappingProxyType(dict(values))


@dataclass(frozen=True, slots=True)
class CityInfo:
    """Station-city context included in many MinuteTemp read responses."""

    id: str = ""
    slug: str = ""
    name: str = ""
    timezone: str = ""


@dataclass(frozen=True, slots=True)
class StationInfo:
    """Station context included in many MinuteTemp read responses."""

    station_id: str = ""
    name: str = ""
    temperature_unit: TemperatureUnit | str = "F"
    uses_nws_climate_day: bool | None = None


@dataclass(frozen=True, slots=True)
class ObservationRecord:
    """OpenAPI-aligned observation record."""

    observation_time: datetime | None = None
    temperature_f: float | None = None
    temperature_c: float | None = None
    dewpoint: float | None = None
    heat_index: float | None = None
    wind_chill: float | None = None
    relative_humidity: float | None = None
    barometric_pressure: float | None = None
    sea_level_pressure: float | None = None
    wind_speed: float | None = None
    wind_direction: float | None = None
    wind_gust: float | None = None
    text_description: str | None = None
    precipitation_1h: float | None = None
    precipitation_3h: float | None = None
    precipitation_6h: float | None = None
    is_locf: bool = False
    is_from_report: bool = False
    report_type: ReportType | None = None
    source_report_id: str | None = None
    temp_min_f: float | None = None
    temp_max_f: float | None = None
    temp_min_c: float | None = None
    temp_max_c: float | None = None


@dataclass(frozen=True, slots=True)
class StationReportRecord:
    """OpenAPI-aligned station report record."""

    report_id: str = ""
    report_revision: int = 0
    report_updated_at: datetime | None = None
    report_type: ReportType | None = None
    report_date: date | None = None
    issuance_time: datetime | None = None
    fetched_at: datetime | None = None
    baseline: bool = False
    provider_available_at: datetime | None = None
    baseline_cached_at: datetime | None = None
    max_temp_f: float | None = None
    max_temp_c: float | None = None
    max_temp_time_utc: datetime | None = None
    min_temp_f: float | None = None
    min_temp_c: float | None = None
    min_temp_time_utc: datetime | None = None
    temp_f: float | None = None
    temp_c: float | None = None


@dataclass(frozen=True, slots=True)
class ReportClockSchedule:
    """Clock-based report schedule definition."""

    basis: ReportScheduleBasis | str = "utc"
    hour: int | None = None
    minute: int | None = None
    utc_hour: int | None = None
    utc_minute: int | None = None
    local_hour: int | None = None
    local_minute: int | None = None
    label: str = ""


@dataclass(frozen=True, slots=True)
class ReportIntervalSchedule:
    """Interval-based report schedule definition."""

    interval_minutes: int
    utc_minute: int | None = None
    local_minute: int | None = None
    label: str = ""


@dataclass(frozen=True, slots=True)
class ReportMultiHourSchedule:
    """Multi-hour report schedule definition."""

    utc_hours: tuple[int, ...] = field(default_factory=tuple)
    local_hours: tuple[int, ...] = field(default_factory=tuple)
    utc_minute: int | None = None
    local_minute: int | None = None
    label: str = ""

    def __post_init__(self) -> None:
        object.__setattr__(self, "utc_hours", _freeze_tuple(self.utc_hours))
        object.__setattr__(self, "local_hours", _freeze_tuple(self.local_hours))


type ReportSchedule = ReportClockSchedule | ReportIntervalSchedule | ReportMultiHourSchedule
type ReportScheduleEntry = ReportSchedule | tuple[ReportSchedule, ...]


def _freeze_report_schedules(
    values: Mapping[str, ReportScheduleEntry],
) -> Mapping[str, ReportScheduleEntry]:
    normalized: dict[str, ReportScheduleEntry] = {}
    for key, value in values.items():
        if isinstance(value, (list, tuple)):
            normalized[key] = _freeze_tuple(value)
        else:
            normalized[key] = value
    return MappingProxyType(normalized)


@dataclass(frozen=True, slots=True)
class HourlyForecastRecord:
    """OpenAPI-aligned hourly forecast point."""

    time: datetime | None = None
    temperature_2m_f: float | None = None
    temperature_2m_c: float | None = None
    apparent_temperature_f: float | None = None
    relative_humidity_2m: float | None = None
    dew_point_2m: float | None = None
    pressure_msl: float | None = None
    wind_speed_10m: float | None = None
    wind_direction_10m: float | None = None
    wind_gusts_10m: float | None = None
    cloud_cover: float | None = None
    precipitation_probability: float | None = None


@dataclass(frozen=True, slots=True)
class ForecastBundleRun:
    """Forecast-run metadata embedded in the latest-forecast response."""

    id: str = ""
    fetched_at: datetime | None = None
    timezone: str = ""
    utc_offset_seconds: int | None = None


@dataclass(frozen=True, slots=True)
class ForecastBundle:
    """One model-specific forecast bundle."""

    model_id: str = ""
    forecast_run: ForecastBundleRun | None = None
    hourly: tuple[HourlyForecastRecord, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        object.__setattr__(self, "hourly", _freeze_tuple(self.hourly))


@dataclass(frozen=True, slots=True)
class OracleModelScoreRecord:
    """OpenAPI-aligned oracle score row."""

    model_id: str = ""
    model_name: str = ""
    is_public: bool | None = None
    high_mae: float | None = None
    low_mae: float | None = None
    high_bias: float | None = None
    low_bias: float | None = None
    combined_mae: float | None = None
    day_count: int | None = None


@dataclass(frozen=True, slots=True)
class OracleScoreData:
    """Typed oracle score response payload."""

    station_id: str = ""
    range_start: date | None = None
    range_end: date | None = None
    days_requested: int | None = None
    all_time: bool | None = None
    score_mode: OracleScoreMode | str = "overall"
    rank_by: OracleRankBy | str = "combined"
    scores: tuple[OracleModelScoreRecord, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        object.__setattr__(self, "scores", _freeze_tuple(self.scores))


@dataclass(frozen=True, slots=True)
class CursorPage:
    """Cursor pagination metadata."""

    limit: int | None = None
    next_cursor: str | None = None


@dataclass(frozen=True, slots=True)
class ForecastRunSummary:
    """OpenAPI-aligned forecast run summary."""

    id: str = ""
    station_id: str = ""
    model_id: str = ""
    forecast_time: datetime | None = None
    fetched_at: datetime | None = None
    timezone: str = ""
    utc_offset_seconds: int | None = None
    data_hash: str = ""


@dataclass(frozen=True, slots=True)
class IpGuardLimits:
    """Per-IP guard limits nested under effective limits."""

    requests_per_second: int | None = None
    burst: int | None = None


@dataclass(frozen=True, slots=True)
class EffectiveLimits:
    """Typed limits response payload."""

    tier: PlanTier | str = "starter"
    requests_per_minute: int | None = None
    daily_max: int | None = None
    max_history_days: int | None = None
    ip_guard: IpGuardLimits | None = None
    rate_limit_remaining: int | None = None
    rate_limit_reset_seconds: int | None = None


@dataclass(frozen=True, slots=True)
class LatestObservationData:
    """Typed latest-observation response payload."""

    city: CityInfo | None = None
    station: StationInfo | None = None
    observation: ObservationRecord | None = None
    daily_high_f: float | None = None
    daily_low_f: float | None = None
    daily_high_c: float | None = None
    daily_low_c: float | None = None
    asos_daily_high_f: float | None = None
    asos_daily_low_f: float | None = None
    asos_daily_high_c: float | None = None
    asos_daily_low_c: float | None = None
    wu_current_temp_f: float | None = None
    wu_current_temp_c: float | None = None
    wu_daily_high_f: float | None = None
    wu_daily_low_f: float | None = None
    wu_daily_high_c: float | None = None
    wu_daily_low_c: float | None = None
    wu_observation_time: datetime | None = None
    wu_fetched_at: datetime | None = None
    temperature_day_mode: TemperatureDayMode | str | None = None
    temperature_day_date: date | None = None
    wu_day_mode: WuDayMode | str | None = None
    wu_day_date: date | None = None


@dataclass(frozen=True, slots=True)
class StationForecastData:
    """Typed latest-forecast response payload."""

    city: CityInfo | None = None
    station: StationInfo | None = None
    forecasts: tuple[ForecastBundle, ...] = field(default_factory=tuple)
    count: int | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "forecasts", _freeze_tuple(self.forecasts))


@dataclass(frozen=True, slots=True)
class LatestReportsData:
    """Typed latest-reports response payload."""

    reports: tuple[StationReportRecord, ...] = field(default_factory=tuple)
    report_schedules: Mapping[str, ReportScheduleEntry] | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "reports", _freeze_tuple(self.reports))
        if self.report_schedules is not None:
            object.__setattr__(self, "report_schedules", _freeze_report_schedules(self.report_schedules))


@dataclass(frozen=True, slots=True)
class StationReportsData:
    """Typed station-reports response payload for filtered non-history reads."""

    reports: tuple[StationReportRecord, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        object.__setattr__(self, "reports", _freeze_tuple(self.reports))


@dataclass(frozen=True, slots=True)
class StationReportHistoryPage:
    """Typed report-history page payload."""

    city: CityInfo | None = None
    station: StationInfo | None = None
    reports: tuple[StationReportRecord, ...] = field(default_factory=tuple)
    count: int | None = None
    page: CursorPage | None = None
    report_schedules: Mapping[str, ReportScheduleEntry] | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "reports", _freeze_tuple(self.reports))
        if self.report_schedules is not None:
            object.__setattr__(self, "report_schedules", _freeze_report_schedules(self.report_schedules))


@dataclass(frozen=True, slots=True)
class ForecastRunsPage:
    """Typed forecast-runs page payload."""

    city: CityInfo | None = None
    station: StationInfo | None = None
    runs: tuple[ForecastRunSummary, ...] = field(default_factory=tuple)
    count: int | None = None
    page: CursorPage | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "runs", _freeze_tuple(self.runs))


@dataclass(frozen=True, slots=True)
class ForecastRunData:
    """Typed single forecast-run detail payload."""

    city: CityInfo | None = None
    station: StationInfo | None = None
    forecast_run: ForecastRunSummary | None = None
    hourly: tuple[HourlyForecastRecord, ...] = field(default_factory=tuple)
    count: int | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "hourly", _freeze_tuple(self.hourly))
